controllers/virtualmachinesnapshot/virtualmachinesnapshot_controller.go

```diff
@@ -158,3 +184,215 @@ func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VirtualMachineSnapshotContext)
+		if err != nil && !errors.Is(err, errParentVMSnapshotNotFound) {
+			return nil, err
+		}
+		if parent != nil {
+			return parent, nil
+		}
+	}
+
+	return nil, errParentVMSnapshotNotFound
+}
+
+func (r *Reconciler) findParentSnapshotHelper(ctx *pkgctx.VirtualMachineSnapshotContext, parentName, target, namespace string) (*vmopv1.VirtualMachineSnapshot, error) {
+	parentVMSnapshot := &vmopv1.VirtualMachineSnapshot{}
+	if err := r.Get(ctx, client.ObjectKey{Name: parentName, Namespace: namespace}, parentVMSnapshot); err != nil {
+		return nil, fmt.Errorf("failed to get parent snapshot %s: %w", parentName, err)
+	}
+
+	for _, child := range parentVMSnapshot.Status.Children {
+		if child.Name == target {
+			ctx.Logger.V(5).Info("found parent snapshot", "parent", parentName, "target", target)
+			return parentVMSnapshot, nil
+		}
+		parent, err := r.findParentSnapshotHelper(ctx, child.Name, target, namespace)
+		if err != nil && !errors.Is(err, errParentVMSnapshotNotFound) {
+			return nil, err
+		}
+		if parent != nil {
+			return parent, nil
+		}
+	}
+	return nil, errParentVMSnapshotNotFound
+}
+
+// Update root snapshots and current snapshot of VM.
+func (r *Reconciler) updateVMStatus(ctx *pkgctx.VirtualMachineSnapshotContext, parentVMSnapshot *vmopv1.VirtualMachineSnapshot) error {
+	ctx.Logger.Info("Updating VM status")
+	if err := r.updateVMCurrentSnapshot(ctx, parentVMSnapshot); err != nil {
+		return err
+	}
+	if err := r.updateVMRootSnapshots(ctx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Set current snapshot of VM to the parent snapshot or to nil.
+func (r *Reconciler) updateVMCurrentSnapshot(ctx *pkgctx.VirtualMachineSnapshotContext, parentVMSnapshot *vmopv1.VirtualMachineSnapshot) error {
+	vm := ctx.VM
+	if vm.Spec.CurrentSnapshot != nil && vm.Spec.CurrentSnapshot.Name != ctx.VirtualMachineSnapshot.Name {
+		ctx.Logger.V(5).Info("VM current snapshot is not the same as the snapshot being deleted, skipping update")
+		return nil
+	}
+	vmPatch := client.MergeFrom(vm.DeepCopy())
+	if parentVMSnapshot != nil {
+		ctx.Logger.V(5).Info("Updating VM current snapshot", "vm", vm.Name, "new current snapshot", parentVMSnapshot.Name)
+		vm.Spec.CurrentSnapshot = vmSnapshotCRToLocalObjectRef(parentVMSnapshot)
+	} else {
+		ctx.Logger.V(5).Info("Updating VM current snapshot", "vm", vm.Name, "new current snapshot", "nil")
+		vm.Spec.CurrentSnapshot = nil
+	}
+	if err := r.Patch(ctx, vm, vmPatch); err != nil {
+		return fmt.Errorf("failed to patch VM %s with current snapshot: %w", vm.Name, err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) updateVMRootSnapshots(ctx *pkgctx.VirtualMachineSnapshotContext) error {
+	ctx.Logger.Info("Updating VM root snapshots")
+	vm := ctx.VM
+	vmSnapshot := ctx.VirtualMachineSnapshot
+	if len(vm.Status.RootSnapshots) == 0 {
+		ctx.Logger.V(5).Info("No root snapshots found for VM, skipping update")
+		return nil
+	}
+	if !slices.Contains(vm.Status.RootSnapshots, *vmSnapshotCRToLocalObjectRef(vmSnapshot)) {
```

Comment:
Should this just be checking by Name, like the removal done later? When we rev the next API version, it isn't totally obvious to me that we'll go update the Status with new APIVersion so this check would fail.